File: 234. Palindrome Linked List/solution.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func isPalindrome(head *ListNode) bool {
    if head == nil || head.Next == nil {
        return true
    }
    if head.Next.Next == nil {
        if head.Val == head.Next.Val {
            return true
        } else {
            return false
        }
    }
    step1 := head
    step2 := head
    for {
        step1 = step1.Next
        step2  = step2.Next.Next
        if step2.Next == nil || step2.Next.Next == nil {
            break
        }
    }
    cur1 := head
    cur2 := reverseLinkedList(step1.Next)
    for {
        if cur1 == nil || cur2 == nil {
            break
        }
        if cur1.Val != cur2.Val {
            return false
        }
        cur1 = cur1.Next
        cur2 = cur2.Next
    }
    return true
}

func reverseLinkedList(head *ListNode) *ListNode {
    if head.Next == nil {
        return head
    }
    p := reverseLinkedList(head.Next)
    head.Next.Next = head
    head.Next = nil
    return p
}

File: 234. Palindrome Linked List/solution.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} head
 * @return {boolean}
 */
var isPalindrome = function(head) {
    if(!head || !head.next) return true
    if(!head.next.next && head.val !== head.next.val) {
        return false
    }
    if(!head.next.next && head.val === head.next.val) {
        return true
    }
    let half = []
    let s1 = head
    let s2 = head
    while(s2 && s2.next) {
        half.push(s1.val)
        s1 = s1.next
        s2 = s2.next.next
    }
    if(s2) {
        s1 = s1.next
    }
    for(let i=0; i<half.length; i++) {
        if(half[half.length-1-i] !== s1.val) {
            return false
        }
        s1 = s1.next
    }
    return true
};